moo.js

```diff
@@ -91,30 +63,73 @@
     return result
   }
 
-  function compileRules(rules, hasStates) {
-    if (!Array.isArray(rules)) rules = objectToRules(rules)
-    var groups = []
-    var parts = []
+  function ruleOptions(name, obj) {
+    if (typeof obj !== 'object' || Array.isArray(obj) || isRegExp(obj)) {
+      obj = { match: obj }
+    }
+    return Object.assign({
+      name: name,
+      lineBreaks: false,
+      pop: false,
+      next: null,
+      push: null,
+    }, obj)
+  }
+
+  function sortRules(rules) {
+    var result = []
     for (var i=0; i<rules.length; i++) {
       var rule = rules[i]
-      var name = rule[0]
-      var obj = rule[1]
 
       // get options
-      if (typeof obj !== 'object' || Array.isArray(obj) || isRegExp(obj)) {
-        obj = { match: obj }
+      var options = ruleOptions(rule[0], rule[1])
+
+      // convert to array
+      var match = options.match
+      if (!Array.isArray(match)) { match = [match] }
+
+      // sort literals by length to ensure longest match
+      var capturingRegexps = []
+      var regexps = []
```

Comment:
`Regexps` (instead of `RegExps`) looks odd to me, but so does `regExps`. Thoughts?